### PR TITLE
[PaddleHackathon No.14]

### PR DESCRIPTION
**Pull request by yangguohao**
### PR types
New features
### PR changes
APIs
### Describe

解决了issues：https://github.com/PaddlePaddle/Paddle/issues/40319
添加了paddle.nn.MultiLabelSoftMarginLoss以及paddle.nn.functional.multi_label_soft_margin_loss的函数
设计文档：https://github.com/PaddlePaddle/community/pull/53
中文文档PR链接:https://github.com/PaddlePaddle/docs/pull/4651

### CODE REVIEW

**Comment by dingjiaweiww** [comment]
在PR被review前请先通过CI噢~

File: python/paddle/fluid/tests/unittests/test_frac_api.py [hunk 1]
```diff
@@ -1,118 +1,118 @@
-#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_frac_api.py]: 这个文件有改动/需要改动吗，没有的话不应该出现在这里。

File: python/paddle/fluid/tests/unittests/test_quantile_and_nanquantile.py [hunk 2]
```diff
@@ -1,268 +1,268 @@
-#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_quantile_and_nanquantile.py]: 这个文件有改动/需要改动吗，没有的话不应该出现在这里。

File: python/paddle/nn/functional/loss.py [hunk 3]
```diff
@@ -2195,3 +2195,45 @@ def hinge_embedding_loss(input, label, margin=1.0, reduction='mean', name=None):
...
         return loss
+
+
+def multi_label_soft_margin_loss(
+    input,
+    label,
+    weight = None,
+    reduction = "mean",
+    name = None):
+    r"""
```
Comment (betterpig) [inline python/paddle/nn/functional/loss.py]: 需要补充英文文档。

File: python/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -2195,3 +2195,45 @@ def hinge_embedding_loss(input, label, margin=1.0, reduction='mean', name=None):
...
+    """
+    if reduction not in ['sum', 'mean', 'none']:
+        raise ValueError(
+            "'reduction' in 'multi_label_soft_margin_loss' should be 'sum', 'mean' or 'none', "
+            "but received {}.".format(reduction))
+
+    if not(input.shape==label.shape):
+        raise ValueError(
+            'The input and label should have some dimension'
```
Comment (betterpig) [inline python/paddle/nn/functional/loss.py]: 1. 单词拼写错了 some->same
2. 加上 `'but received {} != {}'.format(input.shape, label.shape)`

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
+import paddle
+import paddle.fluid as fluid
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: 后续fluid将删除，因此尽量不使用fluid下的API，需要替换为paddle.static.xx，可在[__init__.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/static/__init__.py)中查看static对应的API。若无法替代的，请不要显式使用fluid API，即在开始引入API名称，后面不加路径地调用该API。

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                input_np,
+                label_np,
+                weight_np = None,
+                reduction='mean',
+                functional=False):
+    paddle.enable_static()
+    prog = paddle.static.Program()
+    startup_prog = paddle.static.Program()
+    with paddle.static.program_guard(prog, startup_prog):
+        input = paddle.fluid.data(
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: paddle.fluid.data --> paddle.static.data，下同

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        label = paddle.to_tensor([[0.2, 0.1]], dtype='float32')
+        self.assertRaises(
+            ValueError,
+            paddle.nn.functional.multi_label_soft_margin_loss,
+            input=input,
+            label=label)
+        paddle.enable_static()
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: 最后需要空一行。

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        input = np.random.uniform(0.1, 0.8, size=(20, 30)).astype(np.float64)
+        label =  np.random.randint(0, 2, size=(20, 30)).astype(np.float64)
+        weight = np.random.randint(0, 2, size=(20, 30)).astype(np.float64)
+
+        places = [fluid.CPUPlace()]
+        if fluid.core.is_compiled_with_cuda():
+            places.append(fluid.CUDAPlace(0))
+        reductions = ['sum', 'mean', 'none']
+        for place in places:
+            for reduction in reductions:
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: 前面已经测过place和reduction了，这里可以不用再测了，降低单测执行时间。

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+
+class TestMultiLabelMarginLoss(unittest.TestCase):
+    def test_MultiLabelSoftMarginLoss(self):
+        input = np.random.uniform(0.1, 0.8, size=(20, 30)).astype(np.float64)
+        label = np.random.randint(0, 2, size=(20, 30)).astype(np.float64)
+
+        places = [fluid.CPUPlace()]
+        if fluid.core.is_compiled_with_cuda():
+            places.append(fluid.CUDAPlace(0))
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: 看能否改为 `paddle.static.cpu_places()` 和 `paddle.static.cuda_places(0)`

**Review by betterpig — CHANGES_REQUESTED** [review]
主要问题：
1. 两个单测文件要去掉，如果没修改的话。
2. 尽量将paddle.fluid改为paddle.static

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        if functional:
+            res = call_MultiLabelSoftMarginLoss_functional(input=input,label=label,weight=weight,reduction=reduction)
+        else:
+            res = call_MultiLabelSoftMarginLoss_layer(input=input,label=label,weight=weight,reduction=reduction)
+
+        exe = paddle.static.Executor(place)
+        static_result = exe.run(prog, feed=feed_dict, fetch_list=[res])
+    return static_result
+
+def test_dygraph(place,
```
Comment (betterpig) [inline python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: place在动态图测试中似乎没有体现。

File: python/paddle/fluid/tests/unittests/test_frac_api.py (hunk 1 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/test_frac_api.py]: 这个文件改动吗，能删除吗，请答复一下。

File: python/paddle/fluid/tests/unittests/test_quantile_and_nanquantile.py (hunk 2 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/test_quantile_and_nanquantile.py]: 这个文件改动吗，能删除吗，请答复一下。

**Review by betterpig — CHANGES_REQUESTED** [review]
大部分问题已解决，还需要看下两个无关单测怎么从PR中去除，以及单测中动态图测试函数里并没有用到place。

**Comment by yangguohao** [comment]
那两个无关文件我不知道怎么删除，我并没有改动他们，也都是一样的，不知道为什么判定修改了。

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py (hunk 10 above)
Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: 我修改一下

Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: 这一部分我参考已有的损失函数的test文件，发现place在动态图中都没有体现。

File: python/paddle/fluid/tests/unittests/test_frac_api.py (hunk 1 above)
Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_frac_api.py]: 我用原文件覆盖也判定为修改了，不知道什么原因。

> 这个文件改动吗，能删除吗，请答复一下。

File: python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py (hunk 10 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/test_multilabel_margin_loss.py]: 可能是考虑到在静态图中已经测试了place，那就不用加了。

File: python/paddle/fluid/tests/unittests/test_frac_api.py (hunk 1 above)
Comment (betterpig) [reply python/paddle/fluid/tests/unittests/test_frac_api.py]: 先保留要更改的文件，然后把分支reset成最新的develop分支，再加上修改。

**Review by betterpig — DISMISSED** [review]
LGTM

File: python/paddle/nn/layer/loss.py [hunk 11]
```diff
@@ -1302,3 +1302,69 @@ def forward(self, input, label):
...
+        and output :math:`y` (which is a 2D `Tensor` of target class indices).
+        For each sample in the mini-batch:
+        .. math::
+            \text{loss}(x, y) = \sum_{ij}\frac{\max(0, 1 - (x[y[j]] - x[i]))}{\text{x.size}(0)}
+        where :math:`x \in \left\{0, \; \cdots , \; \text{x.size}(0) - 1\right\}`, \
+        :math:`y \in \left\{0, \; \cdots , \; \text{y.size}(0) - 1\right\}`, \
+        :math:`0 \leq y[j] \leq \text{x.size}(0)-1`, \
+        and :math:`i \neq y[j]` for all :math:`i` and :math:`j`.
+        :math:`y` and :math:`x` must have the same size.
+        Parameters:
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: Parameters、Call parameters、Shape、Returns、Examples 等和前面空一行

File: python/paddle/nn/layer/loss.py [hunk 12]
```diff
@@ -1302,3 +1302,69 @@ def forward(self, input, label):
...
+            label (Tensor): Label tensor containing 1 or -1, the data type is float32 or float64. The shape of label is the same as the shape of input.
+        Shape:
+            input: N-D Tensor, the shape is [N, \*], N is batch size and `\*` means number of classes, available dtype is float32, float64. The sum operationoperates over all the elements.
+            label: N-D Tensor, same shape as the input.
+            output: scalar. If :attr:`reduction` is ``'none'``, then same shape as the input.
+        Returns:
+            Tensor, The tensor variable storing the multi_label_soft_margin_loss of input and label.
+        Examples:
+            .. code-block:: python
+                import paddle
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: import paddle 前要空一行

File: python/paddle/nn/layer/loss.py [hunk 13]
```diff
@@ -1302,3 +1302,69 @@ def forward(self, input, label):
...
+                import paddle.nn as nn
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: 示例代码也要注意空行

File: python/paddle/nn/layer/loss.py [hunk 14]
```diff
@@ -1302,3 +1302,69 @@ def forward(self, input, label):
...
             margin=self.margin,
             name=self.name)
+
+
+class MultiLabelSoftMarginLoss(Layer):
+    r"""Creates a criterion that optimizes a multi-class multi-classification
+        hinge loss (margin-based loss) between input :math:`x` (a 2D mini-batch `Tensor`)
+        and output :math:`y` (which is a 2D `Tensor` of target class indices).
+        For each sample in the mini-batch:
+        .. math::
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: 公式这里有格式问题，前面要空一行

**Comment by luotao1** [comment]
PR_CI_Coverage的历史记录看：https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/builds/6512?module=github/PaddlePaddle/Paddle&pipeline=PR-CI-Coverage&branch=pull/41183(develop) 失败了3次，都是单测超时，可以考虑缩小下单测输入数据大小来减少运行时间

**Comment by yangguohao** [comment]
> PR_CI_Coverage的历史记录看：https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/builds/6512?module=github/PaddlePaddle/Paddle&pipeline=PR-CI-Coverage&branch=pull/41183(develop) 失败了3次，都是单测超时，可以考虑缩小下单测输入数据大小来减少运行时间

ok

File: python/paddle/nn/functional/loss.py [hunk 15]
```diff
@@ -2225,3 +2225,79 @@ def hinge_embedding_loss(input, label, margin=1.0, reduction='mean', name=None):
         return paddle.sum(loss, name=name)
     elif reduction == 'none':
         return loss
+
+def multi_label_soft_margin_loss(
```
Comment (TCChenlong) [inline python/paddle/nn/functional/loss.py]: 这里的英文文档，同下面的API，也需要在相关的位置空行~
这个问题，可以合入后，在提个PR，单独修复文档~
LGTM~~

**Review by TCChenlong — DISMISSED** [review]

File: python/paddle/nn/layer/loss.py [hunk 16]
```diff
@@ -1302,3 +1302,77 @@ def forward(self, input, label):
...
+                "but received {}.".format(reduction))
+        self.weight=weight
+        self.reduction = reduction
+        self.name = name
+
+    def forward(self, input, label):
+        return F.multi_label_soft_margin_loss(
+            input,
+            label,
+            reduction=self.reduction,
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: 这里建议严谨一点， weight 在前，reduction 在后吧

File: python/paddle/nn/layer/loss.py [hunk 17]
```diff
@@ -1302,3 +1302,77 @@ def forward(self, input, label):
...
+        For each sample in the mini-batch:
+        .. math::
+            \text{loss}(x, y) = \sum_{ij}\frac{\max(0, 1 - (x[y[j]] - x[i]))}{\text{x.size}(0)}
+        where :math:`x \in \left\{0, \; \cdots , \; \text{x.size}(0) - 1\right\}`, \
+        :math:`y \in \left\{0, \; \cdots , \; \text{y.size}(0) - 1\right\}`, \
+        :math:`0 \leq y[j] \leq \text{x.size}(0)-1`, \
+        and :math:`i \neq y[j]` for all :math:`i` and :math:`j`.
+        :math:`y` and :math:`x` must have the same size.
+
+        Parameters:
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: 这里少了 weight 的参数说明

File: python/paddle/nn/layer/loss.py [hunk 18]
```diff
@@ -1302,3 +1302,77 @@ def forward(self, input, label):
             reduction=self.reduction,
             margin=self.margin,
             name=self.name)
+
+
+class MultiLabelSoftMarginLoss(Layer):
+    r"""Creates a criterion that optimizes a multi-class multi-classification
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: 英文的这段说明，较中文少了一部分内容，建议语义内容上，与中文保持一致

File: python/paddle/nn/layer/loss.py [hunk 19]
```diff
@@ -1302,3 +1302,77 @@ def forward(self, input, label):
...
+            reduction (str, optional): Indicate how to average the loss by batch_size,
+                    the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
+                    If :attr:`reduction` is ``'none'``, the unreduced loss is returned;
+                    If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
+                    If :attr:`reduction` is ``'sum'``, the summed loss is returned.
+                    Default: ``'mean'``
+            name (str, optional): Name for the operation (optional, default is None).
+                For more information, please refer to :ref:`api_guide_Name`.
+
+        Call parameters:
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: 这个在中文文档中没有体现

File: python/paddle/nn/layer/loss.py [hunk 20]
```diff
@@ -1302,3 +1302,77 @@ def forward(self, input, label):
...
+            input (Tensor): Input tensor, the data type is float32 or float64. Shape is (N, C), where C is number of classes, and if shape is more than 2D, this is (N, C, D1, D2,..., Dk), k >= 1.
+            label (Tensor): Label tensor containing 1 or -1, the data type is float32 or float64. The shape of label is the same as the shape of input.
+
+        Shape:
+            input: N-D Tensor, the shape is [N, \*], N is batch size and `\*` means number of classes, available dtype is float32, float64. The sum operationoperates over all the elements.
+            label: N-D Tensor, same shape as the input.
+            output: scalar. If :attr:`reduction` is ``'none'``, then same shape as the input.
+
+        Returns:
```
Comment (TCChenlong) [inline python/paddle/nn/layer/loss.py]: 这个返回不是 Tensor吧，可以不写返回，写shape就好

File: python/paddle/nn/layer/loss.py (hunk 16 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py]: 已修改

File: python/paddle/nn/layer/loss.py (hunk 17 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py]: 已添加

File: python/paddle/nn/layer/loss.py (hunk 20 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py]: 修改为 A callable object of MultiLabelSoftMarginLoss.

File: python/paddle/nn/functional/loss.py (hunk 15 above)
Comment (yangguohao) [reply python/paddle/nn/functional/loss.py]: 1

File: python/paddle/nn/layer/loss.py (hunk 19 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py]: 已作修改

File: python/paddle/nn/layer/loss.py (hunk 18 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py]: 我看了好像是中文内容少了一些，补充了中文内容

**Review by DDDivano — DISMISSED** [review]

**Review by TCChenlong — DISMISSED** [review]
LGTM

**Comment by TCChenlong** [comment]
@yangguohao 请解决一下 Code Style 问题

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Comment by jeff41404** [comment]
code is fine, but in [RFC](https://github.com/PaddlePaddle/community/pull/53) the description of parameters of `paddle.nn.MultiLabelSoftMarginLoss` is not correct which do not have  input and target

**Review by jeff41404 — APPROVED** [review]
